### DISCUSSION TITLE
Environment variables "leak" into Docker container action. Why? How to prevent?

### DISCUSSION THREAD

**Discussion by madhead** [community/community#25738] [Actions]
<p>Imagine, you have have a custom third-party Docker container action implemented with Java. This actions uses <code>JAVA_HOME</code> environment variable to run it’s logic. Think about Gradle application plugin which creates and executable Java application that relies on that environment variable. Imagine this Docker container action is based on <code>openjdk:17</code>.</p>
<p>This action is really simple, it prints the environment and JDK location:</p>
<pre><code class="lang-bash">#!/usr/bin/env sh

echo "Environment:"
printenv

printf "\n"
echo "JAVA_HOME:"
printenv JAVA_HOME

printf "\n"
echo "Java location:"
command -v java

printf "\n"
echo "Java version:"
java --version
</code></pre>
<p>If you just run this custom action, it would work:</p>
<pre><code class="lang-auto">- uses: madhead/actions-env-leak@main
</code></pre>
<p>The output would be:</p>
<pre><code class="lang-auto">JAVA_HOME:
/usr/java/openjdk-17

Java location:
/usr/java/openjdk-17/bin/java
</code></pre>
<p>Everything is fine.</p>
<p>Now, imagine your workflow also needs Java, but a different version:</p>
<pre><code class="lang-auto">- uses: actions/setup-java@v2
  with:
    distribution: 'adopt'
    java-version: '11'
</code></pre>
<p>Now, if you run the custom action again, it is screwed up:</p>
<pre><code class="lang-bash">JAVA_HOME:
/opt/hostedtoolcache/Java_Adopt_jdk/11.0.11-9/x64

Java location:
/usr/java/openjdk-17/bin/java
</code></pre>
<p>Notice that JAVA_HOME now points to a <strong>wrong location</strong>! This variable just leaked from the workflow! Now, if the custom action relies on it to run the Java application, it would fail, because this path is not valid inside the container!</p>
<p>This is what actually happened with one of my actions. The issue is <a href="https://github.com/madhead/read-java-properties/issues/28" rel="noopener nofollow ugc">discussed here</a>.</p>
<p>I have created a minimal reproducible example <a href="https://github.com/madhead/actions-env-leak" rel="noopener nofollow ugc">here</a>.</p>
<p>Why do the variables leak? Why is this behaviour implicit? Shouldn’t it be explicit? The workflow author may want to override some of the variables, but if should not be the default, shouldn’t it? Can action or workflow authors prevent this leakage from happening right now?</p>
<p>Thanks!</p>

**Comment by madhead** [comment]
<p>Conceptually similar issue: <a href="https://github.com/moby/moby/issues/25" class="inline-onebox" rel="noopener nofollow ugc">Dockerd leaks environment variables into the containers · Issue #25 · moby/moby · GitHub</a></p>

**Comment by madhead** [comment, marked as answer]
<p>This is what a support told me on my request to clarify the behavior:</p>
<blockquote>
<p>Thank you for reaching out to GitHub Support and sorry to hear you’ve experienced this issue.</p>
<p>I’ve reached out to the Actions team and they confirmed that most (if not all) environment variables from the runner are passed through to a container when run to allow access to things like secrets and other custom environment variables.</p>
<p>I’m afraid there are currently no plans to change the behaviour around environment variables in these Actions - sorry about that.</p>
<p>We really appreciate feedback on how we can make GitHub even better and the best way to report any feature requests is directly to us through our feedback form: <a href="https://support.github.com/contact/feedback" class="inline-onebox" rel="noopener nofollow ugc">Share feedback - GitHub Support</a></p>
<p>Our roadmap is now publicly visible, so we recommend that you keep an eye on the <a href="https://github.com/blog" rel="noopener nofollow ugc">GitHub Blog</a> and <a href="https://github.com/github/roadmap" rel="noopener nofollow ugc">roadmap</a> for the latest announcements about new features.</p>
</blockquote>

**Comment by bdw429s** [comment]
@madhead Thanks for posting the response.  Did you ever enter a ticket that can be voted on anywhere?  I'd also love to see a workaround for this.  Errant `JAVA_HOME` env vars in the main runner can wreak havoc on Java-based container actions.

**Reply by madhead** [reply]
No, unfortunately, there is no "public ticket" to vote on. Just drop them a message via https://support.github.com, so they see how numerous we are (two-three peeps, probably, lol) and consider implementing a proper fix.

Neither I have a workaround :disappointed:

**Reply by bdw429s** [reply]
Thank you, I will ping them :)
